docs/guides/optimizers.rst

```diff
@@ -4,7 +4,398 @@
+  :end-before: # --- End
+
+There are some cases when you can't define the gradient analytically. In such cases, it is
+possible to use some approximation methods to compute it. One possibility is to use
+finite differences method as demonstrated in the example below.
+
+.. literalinclude:: ../examples/guides/optimizers.py
+  :language: python
+  :start-after: # finite difference
+  :end-before: # --- End
+
+The :func:`finite_differences_gradient <orquestra.opt.gradients.finite_differences_gradient>`
+function itself returns a function that approximates its input gradient. The second parameter
+controls the step size in the approximation.
+
+The optimizers
+==============
+
+Knowing what functions can be minimized, let us now take a closer look at how to minimize them.
+The objects responsible for the whole process are implementations of the :class:`Optimizer`
+interface. Every Optimizer implements the :meth:`minimize <orquestra.opt.api.Optimizer.minimize>`
+method that accepts the following parameters:
+
+- `cost_function`: the cost function to be optimized.
+- `initial_params`:  the initial guess for the optimal parameters. Its dimension should
+  match the dimension of the first argument of the `cost_function`.
+- `keep_history`: a boolean determining if history of function evaluations should be kept or
+  not. The default is :code:`False`, as storage of evaluation history might eat up a lot of
+  memory.
+
+Of course, optimizers have other methods as well. Those methods are relevant mainly when we
+are implementing our own optimizers, and hence we defer discussing them until later.
+
+But wait, surely some optimization methods require additional parameters and some fine tuning.
+Where do those parameters go, if `minimize` only accepts the arguments listed above?
+That's a very good question. Similarly as we have done with other interfaces, we moved all
+optimizer-specific parametrization into Optimizers'
+:meth:`__init__ <orquestra.opt.api.Optimizer.__init__>` method. This way, once the
+optimizers are constructed, they can be used interchangeably with one another.
+
+Let us illustrate how the optimizers work. In the example below, we construct two optimizers
+and we use them to optimize our sum of squares. We then print the results.
+
+.. literalinclude:: ../examples/guides/optimizers.py
+  :language: python
+  :start-after: # Basic optimization
+  :end-before: # --- End
+
+The results might differ slightly, but here is what we got when running the example.
+
+.. code-block:: text
+
+        history: []
+           nfev: 3
+            nit: 2
+     opt_params: array([0., 0., 0.])
+      opt_value: 0.0
+
+.. code-block:: text
+
+        history: []
+           nfev: None
+            nit: 1000
+     opt_params: array([ 5.11528491e-21, -5.11528491e-21,  5.38660015e-21])
+      opt_value: 8.134774054521418e-41
+
+
+As we can see, once constructed, :class:`ScipyOptimizer <orquestra.opt.optimizers.ScipyOptimizer>`
```

Comment:
Maybe worth mentioning somewhere that the output of minimize is adopted from `scipy`?